Preupload lfs files before commiting

Wauplin opened the pull request:
Related to https://github.com/huggingface/datasets/issues/6257#issuecomment-1736822023 cc @mariosasko @lhoestq.

The goal of this PR is to add the possibility to preupload files before making the `create_commit` call. This is expected to be only a power-user feature meant for users generating huge files on the fly (e.g. basically `datasets` when uploading sharded parquet files). The current problem is that each file has to be committed one by one to avoid out-of-memory issues. This strategy leads to 1. users being rate-limited (too many commits) 2. messy git history.

The solution proposed in this PR is to add a `preupload_lfs_files` methods. It takes a list of `CommitOperationAdd` and upload them to S3 (if LFS). Once all the `CommitOperationAdd` are uploaded, a single `create_commit` call is needed.

To make this work, `CommitOperationAdd` is mutated during the `create_commit` process. Some internal attributes are set to keep track of the status (`_upload_mode`, `_is_uploaded`, `_is_committed`). This is the case no matter if the user uses `preupload_lfs_files` or not.  **This is a breaking change** but I don't expect to break any current workflow. This is only a problem is a user defines a `CommitOperationAdd` object to commit to 2 different repos.

Finally `preupload_lfs_files` also mutates the `CommitOperationAdd` objects **to remove the binary content** from them. This is expected as we want to free up the memory (this is the purpose of the whole PR :smile:).

**Documentation:**
- [guide](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1699/en/guides/upload#preupload-lfs-files-before-commit)
- [preupload_lfs_files](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1699/en/package_reference/hf_api#huggingface_hub.HfApi.preupload_lfs_files) 
- [create_commit](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1699/en/package_reference/hf_api#huggingface_hub.HfApi.create_commit) (added warnings)

---

Here is how it would look like:

```py
>>> from huggingface_hub import CommitOperationAdd, preupload_lfs_files, create_commit, create_repo

>>> repo_id = create_repo("test_preupload").repo_id

>>> operations = [] # List of all `CommitOperationAdd` objects that will be generated
>>> for i in range(5):
...     content = ... # generate binary content
...     addition = CommitOperationAdd(path_in_repo=f"shard_{i}_of_5.bin", path_or_fileobj=content)
...     preupload_lfs_files(repo_id, additions=[addition])
...     operations.append(addition)

# Create commit
>>> create_commit(repo_id, operations=operations, commit_message="Commit all shards")
```

**TODO:**
- [x] release memory after upload
- [x] documentation
- [x] tests
- [x] draft integration with `datasets` (https://github.com/huggingface/datasets/pull/6269)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1699). All of your documentation changes will be reflected on that endpoint.

julien-c commented:
As discussed offline, a good server-centric review of this PR would be required from e.g. @coyotte508 and/or @Pierrci

Wauplin commented:
Thanks for pinging them :+1: 

Apart from the PR review, @lhoestq @mariosasko could you open a branch in `datasets` to test the workflow from this PR? It doesn't have to be merged on your side until we make a merge+release of `huggingface_hub` but I want to be 100% sure we are addressing your problem correctly (as we implement it mostly for the parquet shards use case).

LysandreJik commented on docs/source/en/guides/upload.md:
```diff
@@ -431,6 +431,50 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
+In some cases, you might want to upload huge files to S3 **before** making the commit call. For example, if you are
+committing a dataset in several shards that are generated in-memory, you would need to upload the shards one by one
+to avoid a out-of-memory issue. A solution is to upload each shard as a separate commit on the repo. While being
+perfectly valid, this solution has the drawback of potentially messing the git history by generating a tens of commits.
```
```suggestion
perfectly valid, this solution has the drawback of potentially messing the git history by generating tens of commits.
```

LysandreJik commented on src/huggingface_hub/_commit_api.py:
```diff
@@ -483,14 +493,15 @@ def fetch_upload_modes(
         upload_modes.update(**{file["path"]: file["uploadMode"] for file in preupload_info["files"]})
 
+    # Set upload mode for each addition operation
+    for addition in additions:
+        addition._upload_mode = upload_modes[addition.path_in_repo]
+
     # Empty files cannot be uploaded as LFS (S3 would fail with a 501 Not Implemented)
     # => empty files are uploaded as "regular" to still allow users to commit them.
     for addition in additions:
         if addition.upload_info.size == 0:
-            path = addition.path_in_repo
-            upload_modes[path] = "regular"
-
-    return upload_modes
+            addition._upload_mode = "regular"
```
(nit) This change is a breaking change right? If this is supposed to be an internal method, let's maybe update its name to `_fetch_upload_modes`? It's already in a private `_commit_api.py` file but accessible from `from huggingface_hub.hf_api import fetch_upload_modes`

Wauplin replied:
Yes it's definitely a breaking change. I will rename it to `_fetch_upload_modes`. As you said, it's from a private file and it's undocumented in our docs. I don't expect anyone to use it given how low-level and internal the logic is.

Wauplin replied:
(I've checked in `transformers` & cie first)

LysandreJik commented on src/huggingface_hub/hf_api.py:
````diff
@@ -3392,6 +3400,108 @@ def create_commits_on_pr(
+                Number of concurrent threads for uploading files. Defaults to 5.
+                Setting it to 2 means at most 2 files will be uploaded concurrently.
+
+        Example:
+        ```py
+        ```
````
to complete

Wauplin replied:
Good catch :+1:

LysandreJik commented on docs/source/en/guides/upload.md:
````diff
@@ -431,6 +431,50 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
+
+Here is a simple example illustrating how to pre-upload files:
+
+```py
+>>> from huggingface_hub import CommitOperationAdd, preupload_lfs_files, create_commit, create_repo
+
+>>> repo_id = create_repo("test_preupload").repo_id
+
+>>> operations = [] # List of all `CommitOperationAdd` objects that will be generated
+>>> for i in range(5):
+...     content = ... # generate binary content
+...     addition = CommitOperationAdd(path_in_repo=f"shard_{i}_of_5.bin", path_or_fileobj=content)
+...     preupload_lfs_files(repo_id, additions=[addition])
+...     operations.append(addition)
+
+# Create commit
+>>> create_commit(repo_id, operations=operations, commit_message="Commit all shards")
+```
````
What happens if there is a failure on one of the operations before the commit is created?

Wauplin replied:
If there is a failure in one of the uploads, the script will crash (if not in a try/except). If the script is restarted, already uploaded files will not need to be re-uploaded (`preupload_lfs_files` will be _instant_). However, as long as the `create_commit` operation is not completed, the files are not accessible to anyone and cannot be seen on the repo on the Hub. 

Also there was a question at some point to garbage-collect the untracked files in S3 after some time (after 24h?) cc @Pierrci. I think it's not enabled yet but that would mean that if the user waits too long before creating the commit then it's lost.

lhoestq replied:
Is there a way to check if a certain file has been preuploaded based on its name ?

That would help implementing `push_to_hub` resuming if it crashed mid-way

Wauplin replied:
> Is there a way to check if a certain file has been preuploaded based on its name ? Or it requires the hash ?

It requires the hash unfortunately. Filenames are just convenient aliases saved in git but what matters on S3 are the uniqueness of the files (i.e. based on hash). For context, until the `create_commit` step, the filename is not even provided to the Hub.

lhoestq replied:
Ok I see. I guess we can do a commit every N files to allow resuming from there

Wauplin replied:
@lhoestq Yes indeed.
But in general, what takes the most time in `push_to_hub`? Is it generating + hashing the shards or uploading them? Because if generate+hash takes 5% of the upload time for instance, resuming from a failed `push_to_hub` should still be very fast.

mariosasko replied:
Our current "resuming" logic is to generate the shards and compute their "fingerprint" to check if they are already present in the repo (basically, identical to hashing), so "resuming from a failed `push_to_hub`" should be fine, as this step is pretty fast (I don't remember anyone complaining about it being slow)

Wauplin commented:
Thanks for the review @LysandreJik ! I have addressed your comments. We also got approval from moon-landing side (cc @Pierrci @coyotte508) and @mariosasko create a draft PR to test it in `datasets` so I think we are good to merge if everyone's fine? :)

Pierrci reviewed (commented):
Didn't check the details of the code, but as discussed offline, the logic LGTM 👍

mariosasko commented on docs/source/en/guides/upload.md:
```diff
@@ -431,6 +431,50 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
+
+In some cases, you might want to upload huge files to S3 **before** making the commit call. For example, if you are
+committing a dataset in several shards that are generated in-memory, you would need to upload the shards one by one
+to avoid a out-of-memory issue. A solution is to upload each shard as a separate commit on the repo. While being
```
```suggestion
to avoid an out-of-memory issue. A solution is to upload each shard as a separate commit on the repo. While being
```

mariosasko commented on docs/source/en/guides/upload.md:
```diff
@@ -431,6 +431,50 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
+...     preupload_lfs_files(repo_id, additions=[addition])
+...     operations.append(addition)
+
+# Create commit
+>>> create_commit(repo_id, operations=operations, commit_message="Commit all shards")
```
```suggestion
>>> # Create commit
>>> create_commit(repo_id, operations=operations, commit_message="Commit all shards")
```

mariosasko reviewed (approved):
Thanks! The code looks good.

Some doc improvements:

LysandreJik commented:
LGTM, feel free to merge @Wauplin!

Wauplin commented:
Thanks everyone for the reviews and feedback! I'll merge it now :)
